### DISCUSSION TITLE
Closing a PopUp closes everything

### DISCUSSION THREAD

**Discussion by SB97** [cefsharp/CefSharp#4390] [General]
Hello,

i inherited a project containing CefSharp. v6.0.3 to be specific. Before i can update, i need to hotfix it.

I had the problem that PopUps did not communicate with the main window. I solved in the ILifeSpanHandler with changing 

```C#
        public bool OnBeforePopup(...)
        {
            this.view.InvokeNewWindow(targetUrl);
            newBrowser = null;
            return true;
        }
```

to 

```C#
        public bool OnBeforePopup(...)
        {
            newBrowser = null;
            return false;
        }
```

but now the moment i close a PopUp, OnBeforeClose gets called and everything closes.

**Comment by amaitland** [comment]
> v6.0.3 to be specific.

There's no such version to my knowledge. The version should be in your `packages.config` file.

> but now the moment i close a PopUp, OnBeforeClose gets called and everything closes.

What do you mean `everything closes`? `OnBeforeClose` and `DoClose` should both be called when the popups is closing. A `WM_Close` message should be sent to close the popup window.

Need more information, more detailed code example (no images), etc

**Reply by SB97** [reply]
``` <package id="cef.redist.x64" version="3.3239.1723" targetFramework="net452" />
  <package id="cef.redist.x86" version="3.3239.1723" targetFramework="net452" />
  <package id="CefSharp.Common" version="63.0.3" targetFramework="net452" />
  <package id="CefSharp.WinForms" version="63.0.3" targetFramework="net452" />
```
sry, looks like i made a typo.

With everything i mean that every CefSharp windows is closing. All popups and the main window. 

When having a main windows with a popup, closing that calls:
DoClose
OnBeforeClose
DoClose
OnBeforeClose

```C#
        public bool DoClose(IWebBrowser browserControl, IBrowser browser)
        {
            this.view.InvokeCloseWindow();
            return false;
        }
```

```C#
        public void InvokeCloseWindow()
        {
            this.Close?.Invoke(this, EventArgs.Empty);
        }
```

Thanks to you pointing me to the DoClose and this Thread https://stackoverflow.com/a/59285773/7512261
i changed the DoClose to which seams to solve my problem :)
```C#
        public bool DoClose(IWebBrowser browserControl, IBrowser browser)
        {
            //this.view.InvokeCloseWindow();
            browser.Dispose();
            return false;
        }
```
Is this the intended way to close windows?

**Reply by amaitland** [reply]
You can use `browser.IsPopup` to differentiate between main browser and popups.
